rust/connlib/shared/src/windows.rs

```diff
@@ -16,6 +16,107 @@ pub fn app_local_data_dir() -> Result<PathBuf, Error> {
     Ok(path)
 }
 
+pub mod dns {
+    //! Gives Firezone DNS privilege over other DNS resolvers on the system
+    //!
+    //! This uses NRPT and claims all domains, similar to the `systemd-resolved` control method
+    //! on Linux.
+    //! This allows us to "shadow" DNS resolvers that are configured by the user or DHCP on
+    //! physical interfaces, as long as they don't have any NRPT rules that outrank us.
+    //!
+    //! If Firezone crashes, restarting Firezone and closing it gracefully will resume
+    //! normal DNS operation. The Powershell command to remove the NRPT rule can also be run
+    //! by hand.
+    //!
+    //! The system default resolvers don't need to be reverted because they're never deleted.
+    //!
+    //! <https://superuser.com/a/1752670>
+
+    use anyhow::Result;
+    use std::{os::windows::process::CommandExt, process::Command};
+
+    /// Hides Powershell's console on Windows
+    ///
+    /// <https://stackoverflow.com/questions/59692146/is-it-possible-to-use-the-standard-library-to-spawn-a-process-without-showing-th#60958956>
+    /// Also used for self-elevation
+    const CREATE_NO_WINDOW: u32 = 0x08000000;
+
+    // Unique magic number that we can use to delete our well-known NRPT rule.
+    // Copied from the deep link schema
+    const FZ_MAGIC: &str = "firezone-fd0020211111";
+
+    /// Tells Windows to send all DNS queries to our sentinels
+    ///
+    /// Parameters:
+    /// - `dns_config_string`: Comma-separated IP addresses of DNS servers, e.g. "1.1.1.1,8.8.8.8"
+    pub fn activate(dns_config_string: &str) -> Result<()> {
```

Comment:
I think it's clearer as an interface if we use an `&[IpAddr]` here and format the string within this functions, since these are the ones that should know the formatting

Reply:
Hm true, but it ends up formatting the string twice somewhere IIRC.

What if I make that an opaque type like `NrptRule(String)` (can't remember the exact syntax) which does the formatting on construction and then can be re-used?

Reply:
Found a way to resolve it. The other place the string was used got folded into that same function

Reply:
https://github.com/firezone/firezone/pull/4530/files#diff-c84657356c1a95710239a78a52aaf64ec07bd567c7853f4e66b3bbbf97f47802R53-R57